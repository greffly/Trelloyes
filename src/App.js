import React, { Component } from 'react';
import List from './List';
import './App.css';
import './store.js';

const newRandomCard = () => {
  const id =
    Math.random()
      .toString(36)
      .substring(2, 4) +
    Math.random()
      .toString(36)
      .substring(2, 4);
  return {
    id,
    title: `Random Card ${id}`,
    content: 'lorem ipsum'
  };
};

function omit(obj, keyToOmit) {
  return Object.entries(obj).reduce(
    (newObj, [key, value]) =>
      key === keyToOmit ? newObj : { ...newObj, [key]: value },
    {}
  );
}

class App extends Component {
  static defaultProps = {
    store: {
      lists: [],
      allCards: {}
    }
  };

  state = {
    lists: this.props.store.lists,
    allCards: this.props.store.allCards
  };

  handleAddCard(card, listId) {
    console.log('handle add card', { card });
    let newLists = this.state.lists;
    newLists[listId - 1].cardIds.push(card.id);
    let allCards = this.state.allCards;
    allCards[card.id] = card;
    this.setState({
      lists: newLists,
      allCards: allCards
    });
  }

  handleDeleteCard(cardId) {
    const { lists, allCards } = this.state.store;
    const newLists = lists.map(list => ({
      ...list,
      cardIds: list.cardIds.filter(id => id != cardId)
    }));

    const newCards = omit(allCards, cardId);

    this.setState({
      store: {
        lists: newLists,
        allCards: newCards
      }
    });
  }

  render() {
    return (
      <main className='App'>
        <header className='App-header'>
          <h1>Telloyes!</h1>
        </header>
        <div className='App-list'>
          {this.state.lists.map(list => (
            <List
              key={list.id}
              id={list.id}
              header={list.header}
              cards={list.cardIds.map(id => this.props.store.allCards[id])}
              handleAddCard={listID => {
                this.handleAddCard(newRandomCard(), listID);
              }}
              handleDeleteCard={this.handleDeleteCard}
            />
          ))}
        </div>
      </main>
    );
  }
}

export default App;
