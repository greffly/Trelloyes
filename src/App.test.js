import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import Card from './Card';
import List from './List';
import renderer from 'react-test-renderer';

describe('Telloyes Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Card />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<List />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', () => {
    const tree = renderer
      .create(<div className='App-list' />)
      //somehow add .length to see how many cards are rendered?
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
