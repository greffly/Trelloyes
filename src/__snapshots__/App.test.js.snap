// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Telloyes Component renders the UI as expected 1`] = `
<div
  className="App-list"
/>
`;
